app-modules/research/src/Jobs/Research.php

```diff
@@ -0,0 +1,219 @@
+                    ],
+                    'temperature' => app(AiSettings::class)->temperature,
+                ],
+                'stream' => true,
+                'no_direct_answer' => true,
+            ]);
+
+            $body = $response->getBody();
+
+            $counter = 0;
+            $buffer = '';
+            $eventJoinPattern = "/\r\n\r\n|\n\n|\r\r/";
+
+            while (! $body->eof()) {
+                $buffer .= $body->read(1024);
+
+                while (preg_match($eventJoinPattern, $buffer)) {
+                    $parts = preg_split($eventJoinPattern, $buffer, 2);
+
+                    $rawEvent = $parts[0] ?? '';
+                    $remaining = $parts[1] ?? '';
+
+                    $buffer = $remaining;
+
+                    $matched = preg_match('/(?P<name>[^:]*):?( ?(?P<value>.*))?/', $rawEvent, $matches);
+
+                    if (! $matched) {
+                        continue;
+                    }
+
+                    if (
+                        blank($matches['name']) ||
+                        blank($matches['value'])
+                    ) {
+                        continue;
+                    }
+
+                    if ($matches['name'] !== 'data') {
+                        continue;
+                    }
+
+                    if (! json_validate($matches['value'])) {
+                        continue;
+                    }
+
+                    $data = json_decode($matches['value'], associative: true);
+
+                    $content = $data['choices'][0]['delta']['content'] ?? null;
+
+                    $this->researchRequest->results .= $content;
+
+                    if (blank($content)) {
+                        continue;
+                    }
+
+                    $counter++;
+
+                    if ($counter >= 20) {
+                        $this->researchRequest->save();
+
+                        $counter = 0;
+                    }
+                }
+            }
+
+            dispatch(new RecordTrackedEvent(
+                type: TrackedEventType::AiExchange,
+                occurredAt: now(),
+            ));
+
+            if (blank($this->researchRequest->results)) {
+                $this->researchRequest->results = 'The artificial intelligence service was unavailable. Please try again later.';
+                $this->researchRequest->touch('finished_at');
+
+                return;
+            }
+
+            $this->researchRequest->title = app(AiIntegratedAssistantSettings::class)
+                ->getDefaultModel()
+                ->getService()
+                ->complete(
+                    prompt: $this->researchRequest->results,
+                    content: 'Generate a title for this research, in 5 words or less. Do not respond with any greetings or salutations, and do not include any additional information or context. Just respond with the title:',
+                );
```

Comment:
If this request to the Azure AI fails for any reason it will result in us overwriting the results and considering this a failure.

Perhaps we should wrap this in a try/catch and give a generic title if it failed?

Reply:
Done